component_sdk/python/kfp_component/_base_op.py

```diff
@@ -0,0 +1,213 @@
+
+    def _should_cancel(self):
+        """Checks argo's execution config deadline and decide whether the operation
+        should be cancelled.
+
+        Argo cancels workflow by setting deadline to 0 and sends SIGTERM
+        signal to main container with 10s graceful period.
+        """
+        pod = self._load_pod()
+        if not pod or not pod.metadata or not pod.metadata.annotations:
+            return False
+        
+        argo_execution_config_json = pod.metadata.annotations.get('workflows.argoproj.io/execution', None)
+        if not argo_execution_config_json:
+            return False
+        
+        try:
+            argo_execution_config = json.loads(argo_execution_config_json)
+        except Exception as e:
+            logging.error("Error deserializing argo execution config: {}".format(e))
+            return False
+        
+        deadline_json = argo_execution_config.get('deadline', None)
+        if not deadline_json:
+            return False
+        
+        try:
+            deadline = datetime.strptime(deadline_json, '%Y-%m-%dT%H:%M:%SZ')
+        except Exception as e:
+            logging.error("Error converting deadline string to datetime: {}".format(e))
+            return False
+        
+        return datetime.now() > deadline
+
+    def _stage_states(self):
+        if not self._staging_bucket or not self._staging_path or not self.staging_states:
+            return
+
+        try:
+            bucket = self._storage.get_bucket(self._staging_bucket)
+        except google.cloud.exceptions.NotFound:
+            logging.error('Unable to find staging bucket {}.'.format(self._staging_bucket))
+            return
+ 
+        blob = bucket.blob(self._staging_path)
+
+        try:
+            states_json = json.dumps(self.staging_states)
+        except TypeError as e:
+            logging.error('Unable to dump staging states: {}'.format(self.staging_states))
+            return
+
+        try:
+            blob.upload_from_string(states_json)
+        except google.cloud.exceptions.GoogleCloudError as e:
+            logging.error('Failed to upload staging states: {}'.format(e))
+            return
+
+    def _load_k8s_client(self):
+        config.load_incluster_config()
+        self._v1_core = client.CoreV1Api()
+
+    def _load_argo_metadata(self):
+        # Load argo metadata at start of an OP, as pod might be deleted in case of preemption.
+        pod = self._load_pod()
+        if not pod or not pod.metadata or not pod.metadata.labels or not pod.metadata.annotations:
+            return
+
+        self._argo_workflow_name = pod.metadata.labels.get('workflows.argoproj.io/workflow', None)
+        argo_node_name = pod.metadata.annotations.get('workflows.argoproj.io/node-name', None)
+        if argo_node_name:
+            self._argo_node_name = re.sub(r'\s+\(\d\)', '', argo_node_name)
+
+    def _load_staging_location(self):
+        tmp_location = os.environ.get('KFP_TMP_LOCATION', None)
```

Comment:
What is `tmp_location`?
I think it's better to pass all configuration options through constructor (`__init__`) instead of a function talking directly to the operating system.
Not talking to the OS directly makes it easier to mock and test code.

Reply:
This code is removed.